src/debug_logging.rs

```diff
@@ -0,0 +1,143 @@
+//! Forward log messages to logging webxdc
+use crate::{
+    chat::ChatId,
+    config::Config,
+    context::{Context, DebugLogging},
+    message::{Message, MsgId, Viewtype},
+    param::Param,
+    webxdc::StatusUpdateItem,
+    Event, EventType,
+};
+use async_channel::{self as channel, Receiver};
+use serde_json::json;
+use std::path::PathBuf;
+use tokio::task;
+
+/// Store all information needed to log an event to a webxdc.
+pub struct DebugEventLogData {
+    pub time: i64,
+    pub msg_id: MsgId,
+    pub event: EventType,
+}
+
+/// Creates a loop which forwards all log messages send into the channel to the associated
+/// logging xdc.
+pub async fn debug_logging_loop(context: &Context, events: Receiver<DebugEventLogData>) {
+    while let Ok(DebugEventLogData {
+        time,
+        msg_id,
+        event,
+    }) = events.recv().await
+    {
+        match context
+            .write_status_update_inner(
+                &msg_id,
+                StatusUpdateItem {
+                    payload: json!({
+                        "event": event,
+                        "time": time,
+                    }),
+                    info: None,
+                    summary: None,
+                    document: None,
+                },
+            )
+            .await
+        {
+            Err(err) => {
+                eprintln!("Can't log event to webxdc status update: {:#}", err);
+            }
+            Ok(serial) => {
+                context.events.emit(Event {
+                    id: context.id,
+                    typ: EventType::WebxdcStatusUpdate {
+                        msg_id,
+                        status_update_serial: serial,
+                    },
+                });
+            }
+        }
+    }
+}
+
+/// Set message as new logging webxdc if filename and chat_id fit
+pub async fn maybe_set_logging_xdc(
+    context: &Context,
+    msg: &Message,
+    chat_id: ChatId,
+) -> anyhow::Result<()> {
+    maybe_set_logging_xdc_inner(
+        context,
+        msg.get_viewtype(),
+        chat_id,
+        msg.param.get_path(Param::File, context),
+        msg.get_id(),
+    )
+    .await?;
+
+    Ok(())
+}
+
+/// Set message as new logging webxdc if filename and chat_id fit
+pub async fn maybe_set_logging_xdc_inner(
+    context: &Context,
+    viewtype: Viewtype,
+    chat_id: ChatId,
+    file: anyhow::Result<Option<PathBuf>>,
+    msg_id: MsgId,
+) -> anyhow::Result<()> {
```

Comment:
You could `use anyhow::Result` to not have to write `anyhow::` everywhere.

Reply:
Yeah I know, but I like the explicitness in this case